### PR TITLE
QBD CCC Mapping commit

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

File: apps/mappings/helpers.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from apps.workspaces.models import FyleCredential
+
+from .models import QBDMapping
+
+logger = logging.getLogger(__name__)
+
+def sync_card(workspace_id: int):
+    try:
+        fyle_credentials = FyleCredential.objects.get(workspace_id=workspace_id)
+        platform = Platform(
```
Comment (NileshPant1999) [inline apps/mappings/helpers.py:15]: why are we not using `fyle-integrations-platform-connector` here

File: quickbooks_desktop_api/settings.py [hunk 2]
```diff
@@ -51,13 +51,15 @@
...
+    'django_filters',
 
     # Created Apps
     'apps.users',
     'apps.workspaces',
     'apps.fyle',
     'apps.tasks',
-    'apps.qbd'
+    'apps.qbd',
+    'apps.mappings',
```
Comment (NileshPant1999) [inline quickbooks_desktop_api/settings.py:62]: can remove the extra comma

**Comment by NileshPant1999** [comment]
@DhaaraniCIT you can fix those lint issues that are coming in codefactor

File: apps/mappings/helpers.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	        server_url='{}/platform/v1beta'.format(fyle_credentials.cluster_domain),
+	        token_url=settings.FYLE_TOKEN_URI,
+		    client_id=settings.FYLE_CLIENT_ID,
+		    client_secret=settings.FYLE_CLIENT_SECRET,
+	        refresh_token=fyle_credentials.refresh_token,
+        )
+        query = {
+	        'order': 'updated_at.desc',
```
Comment (Shwetabhk) [inline apps/mappings/helpers.py:23]: wrong indentation

File: apps/mappings/helpers.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	    }
```
Comment (Shwetabhk) [inline apps/mappings/helpers.py:24]: wrong indentation

File: apps/mappings/helpers.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+logger = logging.getLogger(__name__)
+
+def sync_card(workspace_id: int):
+    try:
+        fyle_credentials = FyleCredential.objects.get(workspace_id=workspace_id)
+        platform = Platform(
+	        server_url='{}/platform/v1beta'.format(fyle_credentials.cluster_domain),
+	        token_url=settings.FYLE_TOKEN_URI,
+		    client_id=settings.FYLE_CLIENT_ID,
```
Comment (Shwetabhk) [inline apps/mappings/helpers.py:18]: wrong indentation

File: apps/mappings/helpers.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+		    client_secret=settings.FYLE_CLIENT_SECRET,
```
Comment (Shwetabhk) [inline apps/mappings/helpers.py:19]: wrong indent

File: apps/mappings/helpers.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                        'value': value,
+                        'source_id': card['id'],                 
+                    })
+            if len(card_attributes) > 0:
+                QBDMapping.update_or_create_mapping_objects(card_attributes, workspace_id)
+
+    except FyleCredential.DoesNotExist:
+        logger.info('Fyle credentials not found %s', workspace_id)
+
+    except Exception:
```
Comment (Shwetabhk) [inline apps/mappings/helpers.py:48]: Remove broad exception

File: apps/mappings/helpers.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import logging
```
Comment (Shwetabhk) [inline apps/mappings/helpers.py:1]: The name of this file should be `connector.py` and this should have a class-based structure like all other integrations

Comment (Shwetabhk) [reply apps/mappings/helpers.py:1]:
```python
Class PlatformConnector:
    def __init__(workspace_id):
         initiate connection

    def sync_corporate_cards():
        .....
```

File: apps/mappings/tasks.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+from .helpers import sync_card
+
```
Comment (Shwetabhk) [inline apps/mappings/tasks.py:2]: always add 2 lines between imports and the start of code

File: apps/mappings/tasks.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+def sync_attributes(attribute_type: str, workspace_id: int):
```
Comment (Shwetabhk) [inline apps/mappings/tasks.py:3]: add code comments and function descriptions

File: apps/mappings/views.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+from rest_framework import generics
+from rest_framework.response import Response
+from rest_framework.views import status
+
+from apps.mappings.serializers import QBDMappingSerializer
+from quickbooks_desktop_api.utils import assert_valid, LookupFieldMixin
+
+from .models import QBDMapping
+from .tasks import sync_attributes
+
```
Comment (Shwetabhk) [inline apps/mappings/views.py:11]: add one more line

File: apps/mappings/views.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        sync_attributes(attribute_type, workspace_id)
+
+        return self.list(request, *args, **kwargs)
+
+#mapping stats view
+class QBDMappingStatsView(generics.RetrieveAPIView):
+    """
+    Stats for total mapped and unmapped count for a given attribute type
+    """
+    def get(self, request, *args, **kwargs):
```
Comment (Shwetabhk) [inline apps/mappings/views.py:35]: This code should be in the serializer, please move it
There shouldn't be any logic code in the views

File: apps/mappings/helpers.py (hunk 3 above)
Comment (DhaaraniCIT) [reply apps/mappings/helpers.py:23]: indentations are crt @Shwetabhk , some time in git it shows like this, but the indentations are correct, if it is wrong how it would execute, it will thrown errors

File: apps/mappings/connector.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+from apps.workspaces.models import FyleCredential
+
+from .models import QBDMapping
+
+
+class PlatformConnector:
+
+    def __init__(self, workspace_id: int):
+
+        credentials_object = FyleCredential.objects.get(workspace_id=workspace_id)
```
Comment (NileshPant1999) [inline apps/mappings/connector.py:13]: fyle_credentials instead of credentials_object

File: apps/mappings/connector.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+            server_url='{}/platform/v1beta'.format(credentials_object.cluster_domain),
+            token_url=settings.FYLE_TOKEN_URI,
+            client_id=settings.FYLE_CLIENT_ID,
+            client_secret=settings.FYLE_CLIENT_SECRET,
+            refresh_token=credentials_object.refresh_token,
+        )
+
+        self.workspace_id = workspace_id
+
+        credentials_object.refresh_token = self.platform._Platform__refresh_token
```
Comment (NileshPant1999) [inline apps/mappings/connector.py:25]: @DhaaraniCIT why are we doing this ??

File: apps/mappings/serializers.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
+from rest_framework import serializers
+
+from .models import QBDMapping
+
+
+class QBDMappingSerializer(serializers.ModelSerializer):
+	class Meta:
+		model = QBDMapping
```
Comment (NileshPant1999) [inline apps/mappings/serializers.py:8]: intdentation wrong, maintain consistent 1 tab space

File: apps/mappings/serializers.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+		fields = '__all__'
+		read_only_fields = ('workspace', 'created_at', 'updated_at')
+
+	def create(self, validated_data):
+		workspace_id = self.context['request'].parser_context.get('kwargs').get('workspace_id')
```
Comment (NileshPant1999) [inline apps/mappings/serializers.py:13]: same here maintain consistent indentation

File: apps/mappings/serializers.py [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+		qbd_mapping, _ = QBDMapping.objects.update_or_create(
+			source_id=validated_data['source_id'],
```
Comment (NileshPant1999) [inline apps/mappings/serializers.py:15]: consistant indentation

File: apps/mappings/tasks.py [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+    """
+    Sync Attributes will be called for syncing the attribute's data from the fyle to qbd db
+    """
+
+    qbd_connection = PlatformConnector(workspace_id=workspace_id)
+
+    if attribute_type == 'CORPORATE_CARD':
+        qbd_connection.sync_corporate_card()
```
Comment (NileshPant1999) [inline apps/mappings/tasks.py:12]: i think there are 5 spaces here, maintain consistent 4 spaces

File: apps/mappings/urls.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+from django.urls import path
+
+from .views import (
+    QBDMappingView,
+    QBDMappingStatsView
+)
+
+
+urlpatterns = [
+    path('', QBDMappingView.as_view(), name='QBD-Mapping'),
```
Comment (NileshPant1999) [inline apps/mappings/urls.py:24]: name can be all small

File: quickbooks_desktop_api/tests/settings.py [hunk 20]
```diff
@@ -47,12 +47,14 @@
     'fyle_rest_auth',
     'fyle_accounting_mappings',
     'django_q',
+    'django_filters',
```
Comment (NileshPant1999) [inline quickbooks_desktop_api/tests/settings.py:50]: remove ,

File: quickbooks_desktop_api/settings.py [hunk 21]
```diff
@@ -51,13 +51,15 @@
     'fyle_rest_auth',
     'fyle_accounting_mappings',
     'django_q',
+    'django_filters',
```
Comment (NileshPant1999) [inline quickbooks_desktop_api/settings.py:54]: remove ,

File: apps/workspaces/migrations/0026_merge_20230901_0829.py [hunk 22]
```diff
@@ -0,0 +1,14 @@
+# Generated by Django 3.1.14 on 2023-09-01 08:29
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
Comment (NileshPant1999) [inline apps/workspaces/migrations/0026_merge_20230901_0829.py:6]: @DhaaraniCIT please delete the migrations file and rerun migrations again, this file in uncesessary

File: apps/mappings/serializers.py [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+			workspace_id=workspace_id,
+			defaults={
+				'destination_value': validated_data['destination_value']
```
Comment (NileshPant1999) [inline apps/mappings/serializers.py:18]: please maintain consistent spacing

File: apps/workspaces/migrations/0026_merge_20230901_0829.py (hunk 22 above)
Comment (DhaaraniCIT) [reply apps/workspaces/migrations/0026_merge_20230901_0829.py:6]: when i am deleting this i am getting error @NileshPant1999

File: apps/mappings/serializers.py (hunk 15 above)
Comment (DhaaraniCIT) [reply apps/mappings/serializers.py:8]: The indentations are correct, they seem wrong on git
Screenshot 2023-09-08 at 5 59 03 PM

File: apps/mappings/connector.py [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+            server_url='{}/platform/v1beta'.format(fyle_credentials .cluster_domain),
+            token_url=settings.FYLE_TOKEN_URI,
+            client_id=settings.FYLE_CLIENT_ID,
+            client_secret=settings.FYLE_CLIENT_SECRET,
+            refresh_token=fyle_credentials .refresh_token,
+        )
+
+        self.workspace_id = workspace_id
+
+        fyle_credentials .refresh_token = self.platform._Platform__refresh_token
```
Comment (NileshPant1999) [inline apps/mappings/connector.py:25]: @DhaaraniCIT there is a gap before .
also why is this necessary 😕

File: apps/mappings/connector.py [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+        fyle_credentials .save()
```
Comment (NileshPant1999) [inline apps/mappings/connector.py:26]: same here

File: apps/mappings/connector.py [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+            server_url='{}/platform/v1beta'.format(fyle_credentials.cluster_domain),
+            token_url=settings.FYLE_TOKEN_URI,
+            client_id=settings.FYLE_CLIENT_ID,
+            client_secret=settings.FYLE_CLIENT_SECRET,
+            refresh_token=fyle_credentials.refresh_token,
+        )
+
+        self.workspace_id = workspace_id
+
+        fyle_credentials.refresh_token = self.platform._Platform__refresh_token
```
Comment (NileshPant1999) [inline apps/mappings/connector.py:25]: @ashwin1111 fyle refresh token never expires, why are we doing this, just look at it once

**Review by NileshPant1999 — APPROVED** [review]

File: apps/fyle/tasks.py [hunk 27]
```diff
@@ -94,7 +94,7 @@ def import_credit_card_expenses(workspace_id, accounting_export: AccountingExpor
             state=export_settings.credit_card_expense_state,
             settled_at=last_synced_at if export_settings.credit_card_expense_state == 'PAYMENT_PROCESSING' else None,
             approved_at=last_synced_at if export_settings.credit_card_expense_state == 'APPROVED' else None,
-            filter_credit_expenses=True,
+            filter_credit_expenses=False,
```
Comment (ashwin1111) [inline apps/fyle/tasks.py:97]: @DhaaraniCIT pls confirm with @Shwetabhk if we don't need any other changes for refunds other than this

File: apps/mappings/actions.py [hunk 28]
```diff
@@ -0,0 +1,18 @@
+from .models import QBDMapping
+
+
+def qbd_mapping_stat(source_type: str, workspace_id: int):
+
```
Comment (ashwin1111) [inline apps/mappings/actions.py:5]: doc strings and return type is missing

File: apps/mappings/actions.py [hunk 29]
```diff
@@ -0,0 +1,18 @@
+from .models import QBDMapping
+
+
+def qbd_mapping_stat(source_type: str, workspace_id: int):
```
Comment (ashwin1111) [inline apps/mappings/actions.py:4]: get_qbd_mapping_stats()

File: apps/mappings/connector.py [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+        fyle_credentials.save()
```
Comment (ashwin1111) [inline apps/mappings/connector.py:26]: this is not needed, we do this only for qbo and xero

File: apps/mappings/views.py [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+    ordering_fields = ('source_value',)
+
+    def get(self, request, *args, **kwargs):
+        attribute_type = self.request.query_params.get('attribute_type')
+        workspace_id= self.kwargs['workspace_id']
+        qbd_mapping_count = QBDMapping.objects.filter(
+            workspace_id = workspace_id).count()
+
+        if qbd_mapping_count == 0:
+            sync_attributes(attribute_type, workspace_id)
```
Comment (ashwin1111) [inline apps/mappings/views.py:34]: we shouldn't sync cards at GET call, sync should happen similar to what we do in other integrations, on login we should call that

File: apps/workspaces/migrations/0026_merge_20230901_0829.py (hunk 22 above)
Comment (ashwin1111) [reply apps/workspaces/migrations/0026_merge_20230901_0829.py:6]: what error? pls delete local db and re run migrations, we shouldn't merge any migration

File: apps/workspaces/migrations/0025_auto_20230825_1040.py [hunk 32]
```diff
@@ -0,0 +1,21 @@
...
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='exportsettings',
+            name='credit_card_expense_date',
+            field=models.CharField(choices=[('last_spent_at', 'last_spent_at'),
+            ('spent_at', 'spent_at'),
+            ('posted_at', 'posted_at'),
+            ('created_at', 'created_at')], max_length=255, null=True),
```
Comment (ashwin1111) [inline apps/workspaces/migrations/0025_auto_20230825_1040.py:19]: what's the actual change here? can you delete local db and recreate migrations

File: quickbooks_desktop_api/settings.py (hunk 21 above)
Comment (ashwin1111) [reply quickbooks_desktop_api/settings.py:54]: the array is not ending here, need not remove comma

File: quickbooks_desktop_api/tests/settings.py (hunk 20 above)
Comment (ashwin1111) [reply quickbooks_desktop_api/tests/settings.py:50]: the array is not ending here, need not remove comma

File: apps/mappings/tasks.py [hunk 33]
```diff
@@ -0,0 +1,12 @@
+from .connector import PlatformConnector
+
+
+def sync_attributes(attribute_type: str, workspace_id: int):
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:4]: name it, sync_fyle_dimensions() and move this to fyle/actions.py

**Review by ashwin1111 — APPROVED** [review]

**Review by Shwetabhk — APPROVED** [review]
